packages/compose/src/hooks/use-state-with-history/index.ts

```diff
@@ -0,0 +1,54 @@
+/**
+ * WordPress dependencies
+ */
+import { createUndoManager } from '@wordpress/undo-manager';
+import { useCallback, useRef, useState, useEffect } from '@wordpress/element';
+
+/**
+ * useState with undo/redo history.
+ *
+ * @param initialValue Initial value.
+ * @return Value, setValue, hasUndo, hasRedo, undo, redo.
+ */
+export default function useStateWithHistory< T >( initialValue: T ) {
+	const manager = useRef( createUndoManager() );
+	const [ value, setValue ] = useState( initialValue );
```

Comment:
Why can't the internal state be equal to that of the under manager?

Reply:
I don't understand?

Reply:
Sorry, I thought the state was exposed, but I guess not 

https://github.com/WordPress/gutenberg/blob/40c9f17199b067c75d58f1d918fc26c31b08897f/packages/undo-manager/src/index.js#L75

Reply:
Why no give access to the state though, if it's a low level thing? Then you don't need an intermediate state.

Reply:
I'm not sure, it feels like an internal representation of a history. I think regardless of whether we expose it or not, we need an intermediate state, in some occasions we don't record undo/redo steps. (undoIgnore option)